File: backend/app/core/scopes.py
from typing import Dict, List, TypedDict


class ScopeAction(TypedDict):
    id: str
    name: str
    description: str


class ScopeGroup(TypedDict):
    key: str
    name: str
    description: str
    actions: List[ScopeAction]


SYSTEM_SCOPES: Dict[str, ScopeGroup] = {
    "audit": {
        "key": "audit",
        "name": "Audit Logs",
        "description": "Security events and platform audit logs management.",
        "actions": [
            {
                "id": "audit:read",
                "name": "Read Audit Logs",
                "description": "Allows viewing security audit trails and log events.",
            },
            {
                "id": "audit:write",
                "name": "Write Audit Logs",
                "description": "Allows writing custom audit log entries and exporting audit data.",
            },
        ],
    },
}


def get_all_valid_scopes() -> List[str]:
    """
    """
    flat_scopes: List[str] = []
    for group in SYSTEM_SCOPES.values():
        for action in group["actions"]:
            flat_scopes.append(action["id"])
    return flat_scopes


def get_scopes_registry() -> Dict[str, ScopeGroup]:
    """
    """
    return SYSTEM_SCOPES

File: frontend/app/src/lib/constants/scopes.ts
export interface ScopeAction {
    id: string
    labelKey: string
    descriptionKey?: string
}

export interface ScopeGroup {
    key: string
    labelKey: string
    descriptionKey?: string
    actions: ScopeAction[]
}

export const SYSTEM_SCOPES: ScopeGroup[] = [
    {
        key: 'audit',
        labelKey: 'scopes.audit.title',
        descriptionKey: 'scopes.audit.description',
        actions: [
            {
                id: 'audit:read',
                labelKey: 'scopes.audit.actions.read.label',
                descriptionKey: 'scopes.audit.actions.read.description',
            },
            {
                id: 'audit:write',
                labelKey: 'scopes.audit.actions.write.label',
                descriptionKey: 'scopes.audit.actions.write.description',
            },
        ],
    }
]

export const ALL_SCOPES_FLAT: string[] = SYSTEM_SCOPES.flatMap((group) =>
    group.actions.map((action) => action.id)
)
